Source/NSTextFinder.m

```diff
@@ -0,0 +1,332 @@
+}
+
+- (void) _nextMatch
+{
+  [_finder findStringInTextView: (NSTextView *)_client
+                        forward: YES];
+}
+
+- (void) _previousMatch
+{
+  [_finder findStringInTextView: (NSTextView *)_client
+                        forward: NO];
+}
+
+- (void) _replaceAll
+{
+  [_finder replaceAllInTextView: (NSTextView *)_client
+                onlyInSelection: NO];
+}
+
+- (void) _replace
+{
+  [_finder replaceStringInTextView: (NSTextView *)_client];
+}
+
+- (void) _replaceAndFind
+{
+   [_finder replaceStringInTextView: (NSTextView *)_client];
+   [_finder findStringInTextView: (NSTextView *)_client
+                         forward: YES];
+}
+
+- (void) _setSearchString
+{
+  [_finder takeFindStringFromTextView: (NSTextView *)_client];
+}
+
+- (void) _replaceAllInSelection
+{
+  [_finder replaceAllInTextView: (NSTextView *)_client
+                onlyInSelection: YES];
+}
+
+- (void) _selectAll
+{
+}
+
+- (void) _selectAllInSelection
+{
+}
+
+- (void) _hideFindInterface
+{
+}
+
+- (void) _showReplaceInterface
+{
+  [_finder showFindPanel];
+}
+
+- (void) _hideReplaceInterface
+{
+}
+
+// Validating and performing
+- (void) performTextFinderAction: (id)sender
+{
+  BOOL valid = [self validateUserInterfaceAction: sender];
+  if (valid)
+    {
+      if (_finder == nil)
+        {
+          _finder = [[GSTextFinder alloc] init];
+        }
+    }
+}
+
+- (void) performFindPanelAction: (id)sender
+{
+  BOOL valid = [self validateUserInterfaceAction: sender];
+  if (valid)
+    {
+      [self performAction: [sender tag]];
+    }
+}
+
+- (void) performAction: (NSTextFinderAction)op
+{
+  switch (op)
+    {
+    case NSTextFinderActionShowFindInterface:
+      [self _showFindInterface];
+      break;
+    case NSTextFinderActionNextMatch:
+      [self _nextMatch];
+      break;
+    case NSTextFinderActionPreviousMatch:
+      [self _previousMatch];
+      break;
+    case NSTextFinderActionReplaceAll:
+      [self _replaceAll];
+      break;
+    case NSTextFinderActionReplace:
+      [self _replace];
+      break;
+    case NSTextFinderActionReplaceAndFind:
+      [self _replaceAndFind];
+      break;
+    case NSTextFinderActionSetSearchString:
+      [self _setSearchString];
+      break;
+    case NSTextFinderActionReplaceAllInSelection:
+      [self _replaceAllInSelection];
+      break;
+    case NSTextFinderActionSelectAll:
+      [self _selectAll];
+      break;
+    case NSTextFinderActionSelectAllInSelection:
+      [self _selectAllInSelection];
+      break;
+    case NSTextFinderActionHideFindInterface:
+      [self _hideFindInterface];
+      break;
+    case NSTextFinderActionShowReplaceInterface:
+      [self _showReplaceInterface];
+      break;
+    case NSTextFinderActionHideReplaceInterface:
+      [self _hideReplaceInterface];
+      break;
+    default:
+      NSLog(@"Unknown operation: %ld", op);
+      break;
+    }
+}
+
+- (BOOL) validateUserInterfaceAction: (id<NSValidatedUserInterfaceItem>)item
```

Comment:
The name of this method is validateUserInterfaceItem: The code should look something like this
SEL action = [item action];
if (sel_isEqual(action, @selector(performTextFinderAction:)) || sel_isEqual(action, @selector(performFindPanelAction:)))
    return [self validateAction: [item tag]]

return YES;

Reply:
Done